### PR TITLE
Play 2.7.0-M1 and Scala 2.13.0-M3

### PR DESCRIPTION
**Pull request by marcospereira**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,8 +1,9 @@
 language: scala
 scala:
-- 2.12.3
-- 2.11.11
-sudo: false
+- 2.13.0-M3
+- 2.12.6
+- 2.11.12
+sudo: required
```
Comment (marcospereira) [inline .travis.yml]: Need to change to "required" because of travis-ci/travis-ci#8836.

File: .travis.yml [hunk 2]
```diff
@@ -17,11 +18,11 @@ before_script:
   - export DBUS_SESSION_BUS_ADDRESS=/dev/null
   - export CHROME_BIN=/usr/bin/google-chrome
   - "export DISPLAY=:99.0"
-  - "Xvfb :99 +extension RANDR &"
+  - "Xvfb :99 +extension RANDR 2>/dev/null &"
```
Comment (marcospereira) [inline .travis.yml]: Redirect log to `/dev/null` to avoid too much useless logs in Travis.

File: build.sbt [hunk 3]
```diff
@@ -49,12 +51,11 @@ lazy val `scalatestplus-play` = project
       "org.scalatest" %% "scalatest" % ScalatestVersion,
       "org.seleniumhq.selenium" % "selenium-java" % SeleniumVersion exclude(org = "com.codeborne", name = "phantomjsdriver"),
       "org.seleniumhq.selenium" % "htmlunit-driver" % HtmlUnitVersion,
+      "net.sourceforge.htmlunit" % "htmlunit-cssparser" % CssParserVersion,
       "com.codeborne" % "phantomjsdriver" % PhantomJsDriverVersion,
       "com.typesafe.play" %% "play-test" % PlayVersion,
       "com.typesafe.play" %% "play-ws" % PlayVersion,
-      "com.typesafe.play" %% "play-ahc-ws" % PlayVersion,
-      "com.typesafe.play" %% "play-cache" % PlayVersion % Test,
```
Comment (marcospereira) [inline build.sbt]: This is an ancient dependency. Play cache used to be one of the core dependencies in Play. We split it to a new dependency in 2.6 and had to re-add it here. But it was not necessary. The only reference made to `EhCacheModule` was to disable it.

So we can now remove it.

File: docs/manual/working/scalaGuide/main/tests/code/ScalaFunctionalTestSpec.scala [hunk 4]
```diff
@@ -39,20 +36,19 @@ class ScalaFunctionalTestSpec extends MixedPlaySpec with Results {
       .build()
     // #scalafunctionaltest-application
 
-    val applicationWithRouter = new GuiceApplicationBuilder().router(Router.from {
-      case GET(p"/Bob") =>
-        Action {
-          Ok("Hello Bob") as "text/html; charset=utf-8"
+    val applicationWithRouter = GuiceApplicationBuilder()
+      .appRoutes(app => {
+        case ("GET", "/Bob") => app.injector.instanceOf(classOf[DefaultActionBuilder]) {
```
Comment (marcospereira) [inline docs/manual/working/scalaGuide/main/tests/code/ScalaFunctionalTestSpec.scala]: Not the best idiom. Maybe we need to overload `appRoutes` to have something like:

```scala
def appRoutes(routes: ControllerComponents => PartialFunction[(String, String), Handler]): GuiceApplicationBuilder
```

But for now, this is how we avoid global state while using `GuiceApplicationBuilder` and configuring a `Router`.

File: module/src/main/scala/org/scalatestplus/play/PhantomJSFactory.scala [hunk 5]
```diff
@@ -30,6 +30,7 @@ import org.openqa.selenium.remote.DesiredCapabilities
  * you to fill in the `createWebDriver` method, usually by mixing in one of the `BrowserFactory` subtraits such as
  * `PhantomJSFactory`.
  */
+@deprecated("PhantomJS is no longer actively developed, and support will eventually be dropped", "4.0.0")
```
Comment (marcospereira) [inline module/src/main/scala/org/scalatestplus/play/PhantomJSFactory.scala]: PhantomJS was deprecated by Selenium.

File: scripts/build [hunk 6]
```diff
@@ -14,14 +14,17 @@ if [ $? -ne 0 ]; then
...
 # There are lots of different ways to configure the memory with sbt, and not every sbt installation supports
 # all of them, eg on Travis, sbt doesn't support the -mem option. But JAVA_OPTS should be standard across
 # the board.
 export JAVA_OPTS="-Xmx2G -Xss2M -XX:MaxMetaspaceSize=512M -XX:ReservedCodeCacheSize=192M -Dfile.encoding=UTF-8"
 
+# Enable scripts to configure some additional arguments that will be passed to sbt.
+ADDITIONAL_SBT_ARGUMENTS=${ADDITIONAL_SBT_ARGUMENTS:-""}
+
 build() {
-  "$SBT" --warn ++${SCALA_VERSION} "$@" | grep --line-buffered -v 'Resolving \|Generating \|addons\.\|GConf-WARNING'
```
Comment (marcospereira) [inline scripts/build]: Remove `--warn` so that we can better see tests execution.

File: scripts/test [hunk 7]
```diff
@@ -2,6 +2,10 @@
 
 # Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 
+# 1. webdriver.firefox.logfile=/dev/null disable GeckoDriver logs which are polluting Travis logs
+export ADDITIONAL_SBT_ARGUMENTS="-Dwebdriver.firefox.logfile=/dev/null"
```
Comment (marcospereira) [inline scripts/test]: Reducing useless logs.

File: module/src/main/scala/org/scalatestplus/play/BaseOneAppPerTest.scala [hunk 8]
```diff
@@ -39,9 +39,6 @@ import play.api.test.Helpers
  *       def getConfig(key: String)(implicit app: Application) = app.configuration.getOptional[String](key)
  *       getConfig("ehcacheplugin") mustBe Some("disabled")
  *     }
- *     "start the Application" in {
- *       Play.maybeApplication mustBe Some(app)
```
Comment (richdougherty) [inline module/src/main/scala/org/scalatestplus/play/BaseOneAppPerTest.scala]: :+1:

File: docs/manual/working/scalaGuide/main/tests/code/allbrowserspersuite/ExampleOverrideBrowsersSpec.scala [hunk 9]
```diff
@@ -5,37 +5,43 @@ package scalaguide.tests.scalatest.allbrowserspersuite
...
     Vector(
       FirefoxInfo(firefoxProfile),
-      ChromeInfo
+      ChromeInfo()
     )
 
   // Override app if you need an Application with other than
   // default parameters.
-  override def fakeApplication() =
-    new GuiceApplicationBuilder().disable[EhCacheModule].router(Router.from {
```
Comment (richdougherty) [inline docs/manual/working/scalaGuide/main/tests/code/allbrowserspersuite/ExampleOverrideBrowsersSpec.scala]: It's great that you no longer need to remove the EhCacheModule in every single test!

File: docs/manual/working/scalaGuide/main/tests/code/ScalaFunctionalTestSpec.scala (hunk 4 above)
Comment (richdougherty) [reply docs/manual/working/scalaGuide/main/tests/code/ScalaFunctionalTestSpec.scala]: That's fine for a milestone build, but I agree we'll want something improved.

FYI, in the Play integration I realised there are a lot of different ways that we might want to make an application. I factored out all the recipes into an "ApplicationFactory". Then you can just provide an ApplicationFactory and all the different nice methods are available:

https://github.com/playframework/playframework/blob/084a7ab635934a4ef7ac07e212d044a36a516d85/framework/src/play-integration-test/src/test/scala/play/it/test/ApplicationFactory.scala#L23-L48

File: module/src/test/scala/org/scalatestplus/play/MixedFixtureSpec.scala [hunk 10]
```diff
@@ -16,47 +16,48 @@
...
+    "make the Application available implicitly" in new App(buildApp("foo" -> "bar")) {
       getConfig("foo") mustBe Some("bar")
     }
-    "start the Application" in new App(buildApp("foo" -> "bar", "ehcacheplugin" -> "disabled")) {
-      Play.maybeApplication mustBe Some(app)
-    }
-    "start the Application lazily" in new App(buildApp("foo" -> "bar", "ehcacheplugin" -> "disabled")) {
+    "start the Application lazily" in new App(buildApp("foo" -> "bar")) {
       var count = 0
       class TestApplication extends Application {
```
[the remark this answers is not in the document]
Comment (marcospereira) [reply module/src/test/scala/org/scalatestplus/play/MixedFixtureSpec.scala]: Refactored. :-)

**Comment by richdougherty** [comment]
Looks good :+1:

**Pull request merged**